src/System.Linq/src/System/Linq/Intersect.cs

```diff
@@ -19,8 +19,17 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
             {
                 throw Error.ArgumentNull(nameof(second));
             }
+           
+            var first2 = first.Distinct();
+            var second2 = second.Distinct();
 
-            return IntersectIterator(first, second, null);
+            foreach (TSource element in first2)
+            {
+                if (second2.Contains(element))
+                {
+                    yield return element;
```

Comment:
My previous comment still holds here, except for the possibility of accessing a web service `first.Distinct().Count()` times, etc.